File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n=nums.size();
        vector<vector<int>>ans;
        int i;
        int a;
        sort(nums.begin(),nums.end());
        for(i=0;i<n;i++)
        {
            a = 0 - nums[i];
            int left=i+1;
            int right=n-1;
            while(left<right)
            {
                if(nums[left]+nums[right]==a)
                {
                    vector<int>temp;
                    temp.push_back(nums[left]);
                    temp.push_back(nums[right]);
                    temp.push_back(nums[i]);
                    ans.push_back(temp);
                    while(left<right &&nums[left]==nums[left+1])
                    {
                        left++;
                        
                    }
                    while(left<right &nums[right]==nums[right-1])
                    {
                        
                        right--;
                    }
                    left++;
                    right--;
                }
                else if(nums[left]+nums[right]<a)
                {
                    left++;
                }
                else
                {
                    right--;
                }
            }
            while(i+1<n && nums[i]==nums[i+1])
            {
                i++;
            }
        }
        return ans;
        
    }
};

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    var n = nums.length;
    nums.sort((a,b) => a-b);
    var i;
    var result = [];
    
    for(i=0;i<n;i++){
        var a = nums[i];
        var t = -a;
        var s = i+1;
        var e = n-1;
        while(s<e){
            if(nums[s] + nums[e] == t){
                result.push([a,nums[s],nums[e]]);
                while(s<e && nums[s]== nums[s+1]){
                    s++;
                }
                while(s<e && nums[e]==nums[e-1]){
                    e--;
                }
                 s++;
                e--;
            
            }
            else if(nums[s] + nums[e] < t){
                s++;
            }
            else{
                e--;
            }
        }
        while(i+1<n && nums[i]==nums[i+1]){
            i++;
        }
    }
    return result;
};